### ISSUE TITLE
[Feature]: Only start local ollama server if there isn't already one running

### ISSUE THREAD

**Issue by byteconcepts** [BerriAI/litellm#766]
### The Feature

While trying to get litellm running with autogen and ollama different issues came up.

One word ahead:
I already have a ollama system service run as user ollama which serves not only to localhost, but to all interfaces.
Imho. it would be better for scalability if the ollama host name and port wasn't hardcoded so ollama and litellm could be run on different servers.
 But OK, that's the actual state and at the moment, I can live with it.

When I startet litellm with...
`./venv/bin/litellm --debug --model ollama/mymodel`
...in my project before starting my scripts that use autogen, the already locally running ollama server was disturbed, runner killed or some other issues came up.

I also had some problems with an ollama server started by litellm when some questions couldn't be answered.
I couldn't find the reason and where it crashed, but then sometimes the ollama runner was also killed or hung up.

To avoid all that, I added some code to the file _proxy/proxy_cli.py_ to only start a local ollama server instance by litellm if there isn't one already running.
I do this by checking if there exists a system process which command name is "ollama" and it's first parameter is "serve" using the "psutil" library which should be working for linux, mac and windows.
I intentionally do NOT check the username because in this case the ollama server may be started as root, as a system service with user ollama or a normal user who does it before the litellm proxy is started or in a another project that is run under another user's name.

But I only checked it on linux (for the moment).

Here's the code:
_proxy/proxy_cli.py_:
```
import ....
import psutil
...
...
    if model and "ollama" in model:
        # Only start ollama server if no local running ollama server process is found:
        running_local_ollama_server_found = False
        for proc in psutil.process_iter():
            pinfo = proc.as_dict(attrs=['pid', 'username', 'name', 'cmdline'])
            if len(pinfo['cmdline']) == 2:
                if pinfo['name'] == "ollama" and pinfo['cmdline'][1] == "serve":
                    running_local_ollama_server_found = True
                    # print(str(pinfo['pid']) + " - " + pinfo['username'] + " - " + pinfo['name'] + " - " + pinfo['cmdline'][0] + " - " + pinfo['
                    print("Already running local ollama server found with PID " + str(pinfo['pid'])  + ", skipping start of ollama server")
                    break
        if running_local_ollama_server_found == False:
            print("No local \"ollama serve\" process was found, starting ollama server...")
            run_ollama_serve()
...
```

This works for me and I find it very useful. But maybe I don't have the full overview over the litellm project and miss that I broke something by doing this.

But after adding this piece of code I had no more ollama server problems any more in the play with litellm.

But Imho. this additional code makes litellm more versatile in combination with ollama and different ways to start the ollama server.

It would be nice, if something like this could be added until litellm offers more options to connect to a ollama server than on localhost and the original port. (Maybe I'll work on that, later, too.)

Many thanks in advance for every comment or even adding some code like this to the project.

### Motivation, pitch

I want the local ollama server only be started by litellm if there is not already one running.

**Comment by ghost** [comment]
Hey @byteconcepts does it not work when you do

```
litellm --model ollama/mymodel --api_base <my-api-base> 
```

?

**Comment by ghost** [comment]
https://docs.litellm.ai/docs/simple_proxy#save-model-specific-params-api-base-api-keys-temperature-etc

**Comment by ghost** [comment]
bump on this @byteconcepts

**Comment by byteconcepts** [comment]
Many thanks krrishdholakia,

I really overlooked that parameter; yes, works for me, but I mus admit that I just came home from work and only did a little test.
Very good, thanks for the tip.

But that doesn't make code like mine obsolete when running the ollama server as another user or a system service on the same machine like I do at the moment for my testing.

And now, while the fast test of the --api_base Parameter I mentioned that even if host and port of the --api_base differ from localhost and port 11434, the local ollama server is still started, what it shouldn't be, imho.
So I'll take another look how to also solve this little remaining issue.

But I'm so happy I found your project, it's such a great problem-solver. - Great work!

If I still have overlooked something or haven't thought something to the end, please tell me.

Again, many thanks :-)

**Comment by byteconcepts** [comment]
Allright, I also addressed the case that api_base parameter was defined and shows another address and port then the ollama default ones. - In this case, no local ollama server is needed and probably also not wanted or even not installed.

So here's the complete code addressing all things I mentioned until now in this comment:

```
    if model and "ollama" in model:
        # Only start ollama server if no local running ollama server process is found:
        running_local_ollama_server_found = False
        for proc in psutil.process_iter():
            pinfo = proc.as_dict(attrs=['pid', 'username', 'name', 'cmdline', 'terminal'])
            if len(pinfo['cmdline']) == 2:
                if pinfo['name'] == 'ollama' and pinfo['cmdline'][1] == 'serve':
                    running_local_ollama_server_found = True
                    # print(str(pinfo['pid']) + " - " + pinfo['username'] + " - " + pinfo['name'] + " - " + pinfo['cmdline'][0] + " - " + pinfo['
                    print('LiteLLM: Already running local ollama server found with PID ' + str(pinfo['pid']))
                    print("LiteLLM: Skipping start of local ollama server\n")
                    break

        if running_local_ollama_server_found == False:
            # If the parameter api_base was submitted, check address and port
            # to only start a local ollama server if hostname of IP and port
            # match the ollama defaults.
            start_ollama_server = True
            
            if api_base:
                api_base_parts = re.findall(r'^https?://([^:]*):?(\d*)', api_base)
                # print('[%s]' % ', '.join(map(str, api_base_parts)))

                api_host = api_base_parts[0][0]
                if api_base_parts[0][1]:
                    api_port = api_base_parts[0][1]
                else:
                   api_port = ''

                if (api_host != 'localhost' and api_host != '127.0.0.1') and api_port != '11434':
                    start_ollama_server = False
                if (api_host == 'localhost' or api_host == '127.0.0.1') and api_port == '':
                    start_ollama_server = False
            
            if start_ollama_server == False:
                print('LiteLLM: api_base parameter differs from ollama default address and port.')
                print("LiteLLM: Skipping start of local ollama server\n")
            else:
                print('LiteLLM: No local "ollama serve" process was found.')
                print("LiteLLM: Starting local ollama server...\n")
                run_ollama_serve()
```

As I already mentioned: If I still have overlooked something or haven't thought something to the end, please tell me.

Again, many thanks for the cool project

**Comment by ishaan-jaff** [comment]
Hi @byteconcepts thanks so much for using LiteLLM ! (i'm the co maintainer of LiteLLM) 

I'd love to fix this issue you're facing and learn how we can make the proxy better for you. Would you be open for a quick chat ? 

Sharing my cal for your convenience:
https://calendly.com/ishaan-berri/30min?month=2023-11

**Comment by byteconcepts** [comment]
Hi @ishaan-jaff ,

oh I'm sorry I'm only reading the comment now.

There's no more to tell from my side because I just used the proxy server and that is it.
The only things I mentioned were the things I wrote about here in the issue.

Cause it's a habit of mine to immediately fix issues that hinder my debugging I added the few lines of code shown above to fix the two issues.
For API debugging I simply use something like...

`tcpdump -A -s 0 'tcp port 11434 and (((ip[2:2] - ((ip[0]&0xf)<<2)) - ((tcp[12]&0xf0)>>2)) != 0)' -i lo | tee ~/llm-debug.log`

(Port depending on the target API which communication to debug.)

That was it. - And that's all I know and use (at the moment). - Unfortunately I can't provide any more infos or wishes because now it just"works for me".

But I like to say how pleasantly surprised I was when I saw how engaged you are to almost immediately wanting to solve your users problems and make the application better and better.

Many thanks again for your project and the engagement of you developers. - Great Job!
